temporalio/client.py

```diff
@@ -0,0 +1,935 @@
+            reason: Reason for the termination.
+            run_id: Run ID to terminate. Defaults to using :py:meth:`run_id`. If
+                set to None or there is no :py:meth:`run_id`, this will
+                terminate the latest run for the workflow ID.
+            first_execution_run_id: First run ID that started the workflow. If
+                set, the termination makes sure that the workflow was started
+                with the given run ID.
+
+        TODO(cretz): Raises
+        """
+        if run_id == WorkflowHandle.SELF_RUN_ID:
+            run_id = self._run_id
+        await self._client._impl.terminate_workflow(
+            TerminateWorkflowInput(
+                id=self._id,
+                run_id=run_id,
+                args=args,
+                reason=reason,
+                first_execution_run_id=first_execution_run_id,
+            )
+        )
+
+
+@dataclass
+class StartWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.start_workflow`."""
+
+    workflow: str
+    args: Iterable[Any]
+    id: str
+    task_queue: str
+    execution_timeout: Optional[timedelta]
+    run_timeout: Optional[timedelta]
+    task_timeout: Optional[timedelta]
+    id_reuse_policy: WorkflowIDReusePolicy
+    retry_policy: Optional[temporalio.common.RetryPolicy]
+    cron_schedule: str
+    memo: Optional[Mapping[str, Any]]
+    search_attributes: Optional[Mapping[str, Any]]
+    header: Optional[Mapping[str, Any]]
+    start_signal: Optional[str]
+    start_signal_args: Iterable[Any]
+
+
+@dataclass
+class CancelWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.cancel_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    first_execution_run_id: Optional[str]
+
+
+@dataclass
+class QueryWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.query_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    query: str
+    args: Iterable[Any]
+    reject_condition: Optional[WorkflowQueryRejectCondition]
+
+
+@dataclass
+class SignalWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.signal_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    signal: str
+    args: Iterable[Any]
+
+
+@dataclass
+class TerminateWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.terminate_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    first_execution_run_id: Optional[str]
+    args: Iterable[Any]
+    reason: Optional[str]
+
+
+class Interceptor:
+    """Interceptor for clients.
+
+    This should be extended by any client interceptors.
+    """
+
+    def intercept_client(self, next: "OutboundInterceptor") -> "OutboundInterceptor":
+        """Method called for intercepting a client.
+
+        Args:
+            next: The underlying outbound interceptor this interceptor should
+                delegate to.
+
+        Returns:
+            The new interceptor that will be called for each client call.
+        """
+        return next
+
+
+class OutboundInterceptor:
```

Comment:
Maybe call this `BaseOutboundInterceptor`?

Reply:
I am not sure it has value anymore than a suggestion to call it `BaseTemporalError` would. While some Python classes do this, it's by no means universal (it's not `BaseIntEnum` or `BaseTypedDict`).

Also, I am intentionally allowing this to be instantiated directly (this will be especially true with classes `temporalio.worker.WorkflowOutboundInterceptor`) for the default implementation. I intentionally do not make these `@abstractmethod` or extend `ABC` for that reason.

Reply:
I'm fine with the naming, not sure why we need to allow direct instantiation though for the outbound interceptor though.

Reply:
I don't see much value in having an empty interceptor in the chain TBH but with the inheritance based approach I can see why you'd want to allow this.